meerkat_frontend/src/js/technical/tabletSubmission.js

```diff
@@ -0,0 +1,294 @@
+var current_Year = (new Date()).getFullYear();
+
+function gatherTabletData(locID) {
+
+    // First destroyany pre-existing table.
+    $('#div_tabletCompletness table').bootstrapTable('destroy');
+    $('#div_tabletCompletness table').remove();
+
+    //Clear the arrays
+    columnNameArray = [];
+    tabletDataArray = [];
+
+    //API URLS to get Data
+    buildTableStructure();
+    var locationAPI = "/locations";
+    var week2 = parseInt(week) - 2;
+    var week1 = parseInt(week) - 1;
+
+    var main_week_tot = '/devices/submissions/tot_1?location=' + locID + '&filter=epi_year:eq:' + current_Year + '&filter=epi_week:eq:';
+    var week_tot_Api2 = main_week_tot + week2;
+    var week_tot_Api1 = main_week_tot + week1;
+    var week_tot_Api = main_week_tot + week;
+
+    var main_week_reg = '/devices/submissions/reg_1?location=' + locID + '&filter=epi_year:eq:' + current_Year + '&filter=epi_week:eq:';
+    var week_reg_Api2 = main_week_reg + week2;
+    var week_reg_Api1 = main_week_reg + week1;
+    var week_reg_Api = main_week_reg + week;
+
+    var year_Api_2 = '/devices/submissions/tot_1?location=' + locID + '&filter=date:ge:' + date_builder(2, "year", true) + '&filter=date:le:' + date_builder(2, "year", false);
+    var year_Api_1 = '/devices/submissions/tot_1?location=' + locID + '&filter=date:ge:' + date_builder(1, "year", true) + '&filter=date:le:' + date_builder(1, "year", false);
+    var year_Api = '/devices/submissions/tot_1?location=' + locID + '&filter=date:ge:' + date_builder(0, "year", true) + '&filter=date:le:' + date_builder(0, "year", false);
+
+    var month_Api1 = '/devices/submissions/tot_1?location=' + locID + '&filter=date:ge:' + date_builder(1, "month", true) + '&filter=date:le:' + date_builder(1, "month", false);
+    var month_Api = '/devices/submissions/tot_1?location=' + locID + '&filter=date:ge:' + date_builder(0, "month", true) + '&filter=date:le:' + date_builder(0, "month", false);
+
+
+
+    // The first call will generat the most of the table structure
+    $.getJSON(api_root + week_tot_Api2, function(data) {
+        $.each(data.clinicSubmissions, function(index, value) {
+            fillTabletData(value);
+        });
+
+        //collect all the Json call inside array
+        var deferreds = [
+            $.getJSON(api_root + week_tot_Api1, function(data) {
```

Comment:
If you stored the api URLs in an object, as I suggested above, you would be able to loop through all the keys to create all these deferreds in far fewer lines.

Reply:
Done, its working now